### PR TITLE
Feature/http handler

### PR DESCRIPTION
**Pull request by J0Santos**
Is this how you make a Pull Request? I never made one before :fearful:

### CODE REVIEW

File: includes/http/HTTPHandler.hpp [hunk 1]
```diff
@@ -1,13 +1,20 @@
...
 #include "http/HTTPRequest.hpp"
+#include "http/HTTPResponse.hpp"
 #include "socket/ServerSocket.hpp"
 #include "utils/smt.hpp"
 
 namespace webserv {
 
 void http_handle(smt::shared_ptr<ServerSocket> sock, int client_fd);
+smt::shared_ptr<HTTPResponse>
+    generate_error_response(int code, smt::shared_ptr<ServerBlock> config);
```
Comment (Dutesier) [inline includes/http/HTTPHandler.hpp:14]: looks good, but when possible try to pass shared_ptr by reference. This avoids all of the copy constructors code being ran each time the parameter is passed

File: includes/http/HTTPRequest.hpp [hunk 2]
```diff
@@ -74,4 +74,4 @@ class HTTPRequest {
 
 std::ostream& operator<<(std::ostream& os, HTTPRequest const& req);
 
-#endif
\ No newline at end of file
+#endif
```
Comment (Dutesier) [inline includes/http/HTTPRequest.hpp:77]: Thanks

File: includes/http/HTTPResponse.hpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#ifndef HTTP_RESPONSE_HPP
+#define HTTP_RESPONSE_HPP
+
+#include <map>
+#include <sstream>
+#include <string>
+
+namespace webserv {
+
+struct HTTPResponse {
```
Comment (Dutesier) [inline includes/http/HTTPResponse.hpp:10]: Is this just the temporary version?

File: includes/socket/ServerSocket.hpp [hunk 4]
```diff
@@ -13,8 +13,6 @@ class ServerSocket : public TCPSocket {
         ServerSocket(smt::shared_ptr<ServerBlock> config);
         ~ServerSocket(void);
 
-    private:
```
Comment (Dutesier) [inline includes/socket/ServerSocket.hpp:16]: Should this be public?

File: srcs/http/HTTPResponse.cpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
+#include "http/HTTPResponse.hpp"
+
+namespace webserv {
+
+std::map<int, std::string> HTTPResponse::s_status_map =
+    HTTPResponse::create_status_map();
+
+HTTPResponse::HTTPResponse(int                                status,
+                           std::map<std::string, std::string> header,
```
Comment (Dutesier) [inline srcs/http/HTTPResponse.cpp:9]: Here is another example of why passing arguments by reference is helpful

File: srcs/server/HTTPServer.cpp [hunk 6]
```diff
@@ -26,9 +26,9 @@ void HTTPServer::start(void) {
     std::vector<smt::shared_ptr<ServerBlock> >::iterator it;
     for (it = server.begin(); it != server.end(); it++) {
 
-        smt::shared_ptr<ServerSocket> sock(new ServerSocket(*it));
-        m_socket[sock->sockfd()] = sock;
-        epoll_add(sock->sockfd());
+        ServerSocket* dontUse = new ServerSocket(*it);
+        m_socket[dontUse->sockfd()] = smt::shared_ptr<ServerSocket>(dontUse);
+        epoll_add(dontUse->sockfd());
```
Comment (Dutesier) [inline srcs/server/HTTPServer.cpp:31]: Proceeds to use 😂

File: tests/test_HTTPResponse.cpp [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    ASSERT_NO_THROW(a = new webserv::HTTPResponse(
+                        404, std::map<std::string, std::string>()));
+    ASSERT_NO_THROW(delete a);
+}
+
+TEST(test_HTTPResponse, to_str) {
+
+    // testing without body
+    std::string resp1 = "HTTP/1.1 200 OK\r\nDuarte: gaja\r\nMaria: "
+                        "passivo-agressiva\r\nZe: gaja\r\n\r\n";
```
Comment (Dutesier) [inline tests/test_HTTPResponse.cpp:27]: A definição de passivo agressividade

**Review by Dutesier — COMMENTED** [review]
Great overall, let's try to use a bit more references from here onward

File: includes/http/HTTPResponse.hpp (hunk 3 above)
Comment (J0Santos) [reply includes/http/HTTPResponse.hpp:10]: what do you mean by temporary?

File: srcs/server/HTTPServer.cpp (hunk 6 above)
Comment (J0Santos) [reply srcs/server/HTTPServer.cpp:31]: @mlanca-c

File: tests/test_HTTPResponse.cpp (hunk 7 above)
Comment (mlanca-c) [reply tests/test_HTTPResponse.cpp:27]: Nao fui eu...

File: includes/http/HTTPHandler.hpp (hunk 1 above)
Comment (mlanca-c) [reply includes/http/HTTPHandler.hpp:14]: hadn't though of that, I agree its better that way. I will try to think of it from now on.

File: includes/socket/ServerSocket.hpp (hunk 4 above)
Comment (mlanca-c) [reply includes/socket/ServerSocket.hpp:16]: I made a lot of things public throughout the project with no better reason than it made our life much easier. The code is much more easy to understand as well. Other than that its the fact that I also could not find a better reason not to... But I'm open to suggestions

File: includes/http/HTTPResponse.hpp (hunk 3 above)
Comment (Dutesier) [reply includes/http/HTTPResponse.hpp:10]: temporary
/ˈtɛmp(ə)r(ər)i/
adjective
lasting for only a limited period of time; not permanent.

Comment (J0Santos) [reply includes/http/HTTPResponse.hpp:10]: ah, ok. Thank you for clarifying :smiley: 

We can discuss it, but this was the way I was doing it before I saw the way you did your request.  What would you suggest?

File: includes/socket/ServerSocket.hpp (hunk 4 above)
Comment (Dutesier) [reply includes/socket/ServerSocket.hpp:16]: Keeping it private improves encapsulation. Classes that depend on this one should use it's "interface" and not have to worry about implementation logic. Therefor, if we want to change the implementation in the future, dependent classes won't break as easily. Although I do agree that it makes our live easier, just some food for thought 🤓

File: includes/http/HTTPResponse.hpp (hunk 3 above)
Comment (Dutesier) [reply includes/http/HTTPResponse.hpp:10]: Ahahah, should it be a struct? Is there no need to have private attributes and getters and setters?

File: includes/socket/ServerSocket.hpp (hunk 4 above)
Comment (mlanca-c) [reply includes/socket/ServerSocket.hpp:16]: So what do you suggest? Everything private and getters and setters? Cause if so, @J0Santos has homework :nerd_face:

Comment (Dutesier) [reply includes/socket/ServerSocket.hpp:16]: Ahah that could be the case, but maybe there's no need... I just think when we find the need to access the method or attribute's directly, and therefor "break" encapsulation, that it is a good sign that we might need to do some refactoring in the future

Comment (Dutesier) [reply includes/socket/ServerSocket.hpp:16]: Resolving this conversation as this is easily fixed in the future and is subjective to personal design preferences

File: includes/http/HTTPResponse.hpp (hunk 3 above)
Comment (mlanca-c) [reply includes/http/HTTPResponse.hpp:10]: yes, it should be a struct

Comment (Dutesier) [reply includes/http/HTTPResponse.hpp:10]: Alright, solving conversation then, looks good 👍

**Review by Dutesier — APPROVED** [review]

**Pull request merged**